### DISCUSSION TITLE
Design/Architecture/Concept differences with Qt ?

### DISCUSSION THREAD

**Discussion by linrongbin16** [ratatui/ratatui#1153] [Discussion]
Hi,

ratatui is a TUI framework for terminal, while [Qt](https://doc.qt.io/qt.html) is a GUI framework for desktop.

I hope you could share some thoughts about the [QtWidget class](https://doc.qt.io/qt-6/qtwidgets-index.html) architecture & design? Especially about:

1. [Object ownership & parent-child tree structure](https://doc.qt.io/qt-6/objecttrees.html), a parent widget can have children widgets:
   * Parent owns children ownership, once parent been destroyed, children will be destroyed.
   * Children will always display on top of the parent, and only display inside parent (in geometric shape).
2. [Event system](https://doc.qt.io/qt-6/eventsandfilters.html), user keyboard/mouse event will be handled by callback functions:
   * Parent will first try to calculate whether an event happens inside a child, and dispatch the event to that child if the event happens inside it.
   * Or the parent will handle the event by itself, if there's no child to process the event.
   * Each widget can bind callback functions to process the event.
4. [Signals & Slots](https://doc.qt.io/qt-6/signalsandslots.html), allow any widgets to easily communicate with each other, which is much more flexible than the **event system**.

When I'm studying ratatui framework, I found the `widget` in ratatui is more like a `view`, which is more focused on rendering, there seems no event handlers or other data communications.

> I understand there's big difference between ratatui & Qt, basically they're completely different things:
> * ratatui (TUI) renders with characters on terminal, Qt (GUI) renders directly with OS drawing API.
> * ratatui is written with rust, Qt is written with C++.
> * Qt provides much more other systems than ratatui, such as `QML`, `Qt Quick`, `Qt Test`, etc, even support plugins.
> * And a lot more.
>
> But when looking at the `widget` concept, I found there're many concepts share some commonalities. So I still want to discuss it.

**Comment by joshka** [comment]
1. N/A as Ratatui doesn't track relationships between widgets at all (your app / widgets can choose to do this however). https://ratatui.rs/concepts/rendering/
2. N/A as Ratatui doesn't handle events at all - this is the app's responsibility to interact with the backend (crossterm, termion, termwiz)
3. N/A same

Ratatui mainly handles the move characters to the screen part of an app. The other bits are (currently at least) the app's responsibility.

**Reply by linrongbin16** [reply]
thanks for reply
